src/lib/router.ts

```diff
@@ -100,50 +105,59 @@ export function updateProjectRoute(
   }
 }
 
-export const push = (newRoute: Route): void => {
-  const history = get(historyStore);
+const loadExecutor = mutexExecutor.create();
+
+async function mutexLoad(newRoute: Route): Promise<void> {
+  isLoading.set(true);
+
+  const loadedRoute = await loadExecutor.run(async () => {
+    return loadRoute(newRoute);
+  });
+
+  // Only let the last request through.
+  if (loadedRoute === undefined) {
+    return;
+  }
+
+  activeRouteStore.set(loadedRoute);
+  isLoading.set(false);
+}
 
-  // Limit history to a maximum of 10 steps. We shouldn't be doing more than
-  // one subsequent pop() anyway.
-  historyStore.set([...history, newRoute].slice(-10));
+export async function push(newRoute: Route): Promise<void> {
+  await mutexLoad(newRoute);
 
   const path = import.meta.env.VITE_HASH_ROUTING
     ? "#" + routeToPath(newRoute)
     : routeToPath(newRoute);
 
-  window.history.pushState(newRoute, documentTitle, path);
-};
+  window.history.pushState(newRoute, DOCUMENT_TITLE, path);
+}
 
-export const pop = (): void => {
-  const history = get(historyStore);
-  const newRoute = history.pop();
-  if (newRoute) {
-    historyStore.set(history);
-    window.history.back();
-  }
-};
+export function pop(): void {
+  window.history.back();
+}
 
-export function replace(newRoute: Route): void {
-  historyStore.set([newRoute]);
+export async function replace(newRoute: Route): Promise<void> {
```

Comment:
No need to export this.

Reply:
280c51187a16c5831f8527e03f89dd38744092cf